src/radical/pilot/agent/lm/jsrun.py

```diff
@@ -27,85 +27,82 @@ def _configure(self):
         ----------
         slots : List of dictionaries.
+
             The slots that the unit will be placed. A slot has the following
             format:
-            {"cores_per_node": 16,
-             "lfs_per_node": {"size": 0, "path": "/dev/null"},
-             "nodes": [{"lfs": {"path": "/dev/null", "size": 0},
-                        "core_map": [[0]],
-                        "name": "a",
-                        "gpu_map": [],
-                        "uid": 1}],
-             "lm_info": "INFO",
-             "gpus_per_node": 6
+
+            {"nodes"         : [{"name"    : "a",
+                                 "uid"     : 1,
+                                 "gpu_map" : [],
+                                 "core_map": [[0]],
+                                 "lfs"     : {"path": "/dev/null", "size": 0}
+                                }],
+             "cores_per_node": 16,
+             "gpus_per_node" : 6
+             "lfs_per_node"  : {"size": 0, "path": "/dev/null"},
+             "lm_info"       : "INFO",
             }
 
-        id : str
-            The ID of the unit.
-        
-        sandbox : str
-            The unit's sandbox path
+        uid     : unit ID (string)
+        sandbox : unit sandbox (string)
+
         """
 
-        node_ids = list()
-        core_maps = list()
-        gpu_maps = list()
-        for node in slots['nodes']:
-            node_ids += [node['uid']] * len(node['core_map'])
-        core_maps = [core_map for core_map in node['core_map'] for node in slots['nodes']]
-        gpu_maps = [gpu_map for gpu_map in node['gpu_map'] for node in slots['nodes']]
-        rs_id = 0
+        rs_id  = 0
         rs_str = ''
-        if len(gpu_maps):
-            for node_id,core_ids,gpu_ids in zip(node_ids, core_maps, gpu_maps):
-                rs_str += 'RS %d : { ' % rs_id
-                rs_str += 'host: %d, ' % node_id
-                rs_str += 'cpu: %s, ' % ' '.join(map(str,core_ids))
-                rs_str += 'gpu: %s}\n' % ' '.join(map(str,gpu_ids))
-                rs_id += 1
-        else:
-            for node_id,core_ids in zip(node_ids, core_maps):
-                rs_str += 'RS %d : { ' % rs_id
-                rs_str += 'host: %d, ' % node_id
-                rs_str += 'cpu: %s ' % ' '.join(map(str,core_ids))
-                rs_str += '}\n'
-                rs_id += 1
-        
-        rs_name = sandbox + '/rs_layout_cu_%s' % cuid
-        with open(rs_name, 'w') as rs_file:
-            rs_file.write(rs_str)      
+
+        for node in slots['nodes']:
+
+            cores = ' '.join([str(core_set[0]) for core_set
+                                               in  node['core_map']])
+            gpus  = ' '.join([str(gpu_set[0])  for gpu_set
+                                               in  node['gpu_map']])
+
+            rs_str           += 'RS %d: {'  % rs_id
+            rs_str           += ' host: %d' % node['uid']
+            if cores: rs_str += ' cpu: %s'  % cores
```

Comment:
Not sure if we should have an 'if' condition for cores. Two reasons:

* I am not sure if jsrun will work with a layout file which does not specify a cpu to use.
* Under our constraint that using gpus requires at least one cpu, we will always request at least one cpu.

Reply:
True for summit - I am not sure for other systems.

The decision if GPUs processes require a CPU core is made in the scsheduler,  and it is configurable for both options.   So the executor should not reason about it, just enact it.

Reply:
Ah yes I forgot about the over subscribe config. Thanks.